src/EFCore.PG/Storage/Internal/Mapping/NpgsqlArrayArrayTypeMapping.cs

```diff
@@ -0,0 +1,208 @@
+
+            public override Type Type => typeof(TElem[]);
+
+            static bool Compare(TElem[] a, TElem[] b, ValueComparer<TElem> elementComparer)
+            {
+                if (a.Length != b.Length)
+                    return false;
+
+                // Note: the following currently boxes every element access because ValueComparer isn't really
+                // generic (see https://github.com/aspnet/EntityFrameworkCore/issues/11072)
+                for (var i = 0; i < a.Length; i++)
+                    if (!elementComparer.Equals(a[i], b[i]))
+                        return false;
+
+                return true;
+            }
+
+            static TElem[] Snapshot(TElem[] source, ValueComparer<TElem> elementComparer)
+            {
+                if (source == null)
+                    return null;
+
+                var snapshot = new TElem[source.Length];
+                // Note: the following currently boxes every element access because ValueComparer isn't really
+                // generic (see https://github.com/aspnet/EntityFrameworkCore/issues/11072)
+                for (var i = 0; i < source.Length; i++)
+                    snapshot[i] = elementComparer.Snapshot(source[i]);
+                return snapshot;
+            }
+        }
+
+        class SingleDimComparerWithIEquatable<TElem> : ValueComparer<TElem[]>
+            where TElem : IEquatable<TElem>
+        {
+            public SingleDimComparerWithIEquatable() : base(
+                (a, b) => Compare(a, b),
+                o => o.GetHashCode(), // TODO: Need to get hash code of elements...
+                source => DoSnapshot(source)) {}
+
+            public override Type Type => typeof(TElem[]);
+
+            static bool Compare(TElem[] a, TElem[] b)
+            {
+                if (a.Length != b.Length)
+                    return false;
+
+                for (var i = 0; i < a.Length; i++)
+                {
+                    var elem1 = a[i];
+                    var elem2 = b[i];
+                    // Note: the following null checks are elided if TElem is a value type
+                    if (elem1 == null)
+                    {
+                        if (elem2 == null)
+                            continue;
+                        return false;
+                    }
+
+                    if (!elem1.Equals(elem2))
+                        return false;
+                }
+
+                return true;
+            }
+
+            static TElem[] DoSnapshot(TElem[] source)
+            {
+                if (source == null)
+                    return null;
+                var snapshot = new TElem[source.Length];
+                for (var i = 0; i < source.Length; i++)
+                    snapshot[i] = source[i];
+                return snapshot;
+            }
+        }
+
+        class SingleDimComparerWithEquals<TElem> : ValueComparer<TElem[]>
```

Comment:
Why not to have only one comparer which internally will call `EqualityComparer<TElem>.Defeault.Equals`? The JIT can optimize some cases and inline method bodies (see dotnet/coreclr#14125).

Reply:
I agree there's probably some simplification to be done here, but let's do this in another PR, OK? This one is mostly about list translations and I don't want to go too far with other stuff etc...